### PR TITLE
Calling poseidon2-gl from Rust

### PR DESCRIPTION
**Pull request by lvella**

### CODE REVIEW

File: riscv/tests/riscv_data/poseidon_gl_via_coprocessor/src/main.rs [hunk 1]
```diff
@@ -1,53 +1,36 @@
...
-        0,
-        0,
-        0,
-        0,
-    ];
-    let h = poseidon_gl_unsafe(&mut i);
-    assert_eq!(h[0], 4330397376401421145);
-    assert_eq!(h[1], 14124799381142128323);
-    assert_eq!(h[2], 8742572140681234676);
-    assert_eq!(h[3], 14345658006221440202);
```
Comment (lvella) [inline riscv/tests/riscv_data/poseidon_gl_via_coprocessor/src/main.rs:48]: This test doesn't make sense. With the new `Goldilocks` type, the nonsenseness is explicit.

The idea is to rely on the wraparound behavior of a value >= PRIME inside the PoseidonGL machine to make this input be equivalent to all zeros.

It relies too much on implementation details. It should be UB to pass a value that doesn't fit into Goldilocks to PoseidonGL.

File: riscv/src/large_field/runtime.rs [hunk 2]
```diff
@@ -362,6 +365,37 @@ impl Runtime {
         self
     }
 
+    fn with_poseidon2(mut self) -> Self {
+        self.add_submachine(
+            "std::machines::hash::poseidon2_gl::Poseidon2GL",
+            None,
+            "poseidon2_gl",
+            vec!["memory", "split_gl"],
+            [r#"instr poseidon2_gl X, Y
```
Comment (leonardoalt) [inline riscv/src/large_field/runtime.rs:374]:
```suggestion
            [r#"instr poseidon2 X, Y
```

File: riscv/src/large_field/runtime.rs [hunk 3]
```diff
@@ -362,6 +365,37 @@ impl Runtime {
...
+                    // make sure tmp1_col and tmp2_col are aligned memory addresses
+                    tmp3_col * 4 = tmp1_col,
+                    tmp4_col * 4 = tmp2_col,
+                    // make sure the factors fit in 32 bits
+                    tmp3_col = X_b1 + X_b2 * 0x100 + X_b3 * 0x10000 + X_b4 * 0x1000000,
+                    tmp4_col = Y_b5 + Y_b6 * 0x100 + Y_b7 * 0x10000 + Y_b8 * 0x1000000
+                }
+            "#],
+            0,
+            vec!["poseidon2_gl 0, 0;"],
```
Comment (leonardoalt) [inline riscv/src/large_field/runtime.rs:388]:
```suggestion
            vec!["poseidon2 0, 0;"],
```

File: riscv/src/large_field/runtime.rs [hunk 4]
```diff
@@ -362,6 +365,37 @@ impl Runtime {
...
+        );
+
+        // The poseidon2 syscall has input address passed on x10 and output address passed on x11,
+        // they can overlap.
+        let implementation = std::iter::once("poseidon2_gl 10, 11;".to_string());
```
Comment (leonardoalt) [inline riscv/src/large_field/runtime.rs:393]:
```suggestion
        let implementation = std::iter::once("poseidon2 10, 11;".to_string());
```

**Review by leonardoalt — COMMENTED** [review]
Nice!!

File: riscv-runtime/src/goldilocks.rs [hunk 5]
```diff
@@ -0,0 +1,35 @@
+pub const PRIME: u64 = 0xffffffff00000001;
+
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq)]
+pub struct Goldilocks(
+    // TODO: maybe represent this as a u32, which ends up as a full word in Powdr.
```
[the remark this answers is not in the document]
Comment (lvella) [reply riscv-runtime/src/goldilocks.rs:6]: Each RISC-V processor word is represented as a field element, so it does fit a u32, albeit with an invalid value for a real u32. Of course, only works on Goldilocks or bigger prime fields.

File: riscv-runtime/src/hash.rs [hunk 6]
```diff
@@ -1,35 +1,43 @@
...
 /// "cap", the return value is placed in data[..4] and the reference to this
 /// sub-array is returned.
-///
-/// This is unsafe because it does not check if the u64 elements fit the
-/// Goldilocks field.
-pub fn poseidon_gl_unsafe(data: &mut [u64; 12]) -> &[u64; 4] {
+pub fn poseidon_gl(data: &mut [Goldilocks; 12]) -> &[Goldilocks; 4] {
     unsafe {
-        asm!("ecall", in("a0") data as *mut [u64; 12], in("t0") u32::from(Syscall::PoseidonGL));
+        asm!("ecall", in("a0") data as *mut _ as *mut [u64; 12], in("t0") u32::from(Syscall::PoseidonGL));
```
[the remark this answers is not in the document]
Comment (lvella) [reply riscv-runtime/src/hash.rs]: fixed

**Comment by lvella** [comment]
No matter how you implement keccack, the same input yields the same output.

Not poseidon. Each version of poseidon 1 or 2 is dependent on the specific ground field it is defined.

Poseidon2-GL is a different function from Poseidon2-BB, so I am against the change in the name.

File: riscv-syscalls/src/lib.rs [hunk 7]
```diff
@@ -55,4 +55,5 @@ syscalls!(
     (7, KeccakF, "keccakf"),
     (8, Mod256, "mod_256"),
     (9, Halt, "halt"),
+    (10, Poseidon2GL, "poseidon2_gl"),
```
[the remark this answers is not in the document]
Comment (lvella) [reply riscv-syscalls/src/lib.rs:58]: No matter how you implement keccack, the same input yields the same output.

Not poseidon. Each version of poseidon 1 or 2 is dependent on the specific ground field it is defined.

Poseidon2-GL is a different function from Poseidon2-BB, so I am against the change in the name.

File: riscv-runtime/src/goldilocks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lvella) [reply riscv-runtime/src/goldilocks.rs:6]: But we probably should panic if `struct Goldilocks` or the `poseidon*_gl` is used in a field other than goldilocks.

File: riscv-syscalls/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (leonardoalt) [reply riscv-syscalls/src/lib.rs:58]: Ok so what do we call here: https://github.com/powdr-labs/powdr/pull/2046/files#diff-0af01215cf3a71d67dbd4488c361437a9a8c9cd217722230fe03f926cc15ff32R17

File: riscv-runtime/src/goldilocks.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (leonardoalt) [reply riscv-runtime/src/goldilocks.rs:6]: > Each RISC-V processor word is represented as a field element, so it does fit a u32, albeit with an invalid values for a real u32. Of course, only works on Goldilocks or bigger prime fields.

yea agree. any ideas how to do that?

[the remark this answers is not in the document]
Comment (lvella) [reply riscv-runtime/src/goldilocks.rs:6]: We could have ecalls for "MergeFe" and "SplitFe". We already have an implementation for "SplitFe" in `SplitGL`.

[the remark this answers is not in the document]
Comment (leonardoalt) [reply riscv-runtime/src/goldilocks.rs:6]: What I meant is the part where we fool the compiler to let us compile to a single word but put whatever value in it. I guess asm should work

[the remark this answers is not in the document]
Comment (lvella) [reply riscv-runtime/src/goldilocks.rs:6]: Yes, asm. Any ecall requires asm.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**